### PR TITLE
fix(#1478) Button with `display: none` breaks keyboard navigation in <FocusScope contain>

### LINKED ISSUE #1478: Button with `display: none` breaks keyboard navigation in <FocusScope contain>
# 🐛 Bug Report 

Keyboard navigation breaks in a `<FocusScope>` with `contain` set when there's a button with `display: none` inside the focus scope.

## 🤔 Expected Behavior

In a focus scope with three buttons, where the middle one is hidden with `display: none`, you should be able to navigate to the first and the last button with your keyboard (navigation with tab key).

## 😯 Current Behavior

In a focus scope with three buttons, where the middle one is hidden with `display: none`, you get stuck on the first button unable to reach the last button when navigating with keyboard (navigation with tab key).

## 🔦 Context

We noticed this issue when using a modal component we've built with react-aria where we hide a button with `display: none` based on a media query. It's not possible to navigate past the hidden button, you can only go as far as the last button in the tab order before the hidden one.

## 💻 Code Sample

I have recreated the issue [here](https://codesandbox.io/s/festive-nightingale-dd8fs) in a simpler version without the modal.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-aria/focus   | 3.2.3
| Browser          | Safari 14.0.2 (16610.3.7.1.9)
| Operating System | macOS 11.1

### PR DESCRIPTION
**Pull request by majornista**
Filter elements returned by FocusScope.getFocusableElementsInScope to only include visible and displayed elements.

Closes #1478 https://github.com/adobe/react-spectrum/issues/1570

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue 1478](https://github.com/adobe/react-spectrum/issues/1478).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Behavior is covered by unit tests: `yarn test`, and can be replicated using something like this: 

```tsx
function Example() {
  let [isOpen, setOpen] = React.useState(false);
  return (
    <>
      <button onClick={() => setOpen(true)}>Open</button>
      {isOpen &&
        <FocusScope contain restoreFocus autoFocus>
          <input aria-label="First input" placeholder="First input" />
          <input aria-label="Hidden input 1" placeholder="Hidden input 1" hidden />
          <input aria-label="Hidden input 2" placeholder="Hidden input 2" style={{display: 'none'}} />
          <input aria-label="Hidden input 3" placeholder="Hidden input 3" style={{visibility: 'hidden'}} />
          <input aria-label="Hidden input 4" placeholder="Hidden input 4" style={{visibility: 'collapse'}} />
          <input aria-label="Second input" placeholder="Second input" />
          <button onClick={() => setOpen(false)}>Close</button>
        </FocusScope>
      }
    </>
  );
}
```
## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

File: packages/@react-aria/utils/src/isElementVisible.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+function isAttributeVisible(element: Element, childElement?: Element) {
+  return (
+    !element.hasAttribute('hidden') &&
+    (element.nodeName === 'DETAILS' &&
+      childElement &&
+      childElement.nodeName !== 'SUMMARY'
+      ? element.hasAttribute('open')
+      : true)
+  );
+}
```
Comment (LFDanLu) [inline packages/@react-aria/utils/src/isElementVisible.ts:49]: I know this was sourced from jest-dom and that our selector filtering/isStyleVisible will ensure that elements with `hidden` aren't included but its kinda weird that this resolves to true if the element has `hidden` as an attribute

**Review by LFDanLu — COMMENTED** [review]
I'm fine with the approach taken here, did you have any luck with/investigate the treewalker approach?

File: packages/@react-aria/utils/src/isElementVisible.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    return false;
+  }
+
+  let {display, visibility} = element.style;
+
+  let isVisible = (
+    display !== 'none' &&
+    visibility !== 'hidden' &&
+    visibility !== 'collapse'
+  );
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/utils/src/isElementVisible.ts:24]: Trying to avoid DOM reflow if it is unnecessary. `element.style` doesn't just reads the `style` prop, while `getComputedStyles` redraws.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 3]
```diff
@@ -179,7 +179,10 @@ function getFocusableElementsInScope(scope: HTMLElement[], opts: FocusManagerOpt
     }
     res.push(...Array.from(node.querySelectorAll(selector)));
   }
-  return res;
+
+  // Filter to include only displayed and visible elements in the DOM,
+  // which accounts for elements hidden based on responsive layout.
+  return res.filter(node => isElementVisible(node));
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:185]: Not at this point, I don't think. It only matters when looping up through ancestors. Here, we're filtering focusable elements already selected by `querySelectorAll`, but we need to confirm that the elements returned are actually visible when rendered with CSS, which could potentially hide elements based on media queries, etc.

File: packages/@react-aria/utils/src/isElementVisible.ts (hunk 1 above)
Comment (majornista) [reply packages/@react-aria/utils/src/isElementVisible.ts:49]: Wouldn't that case be covered by `!element.hasAttribute('hidden') &&` at line 42?

**Comment by majornista** [comment]
> I'm fine with the approach taken here, did you have any luck with/investigate the treewalker approach?

I looked into it, but the `scope` in FocusScope is an array of child elements contained within the FocusScope component, and doesn't really seem connected to a DOM container, so this seemed like a simpler solution.

File: packages/@react-aria/utils/src/isElementVisible.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LFDanLu) [reply packages/@react-aria/utils/src/isElementVisible.ts:24]: Gotcha, thanks

File: packages/@react-aria/utils/src/isElementVisible.ts (hunk 1 above)
Comment (LFDanLu) [reply packages/@react-aria/utils/src/isElementVisible.ts:49]: That then resolves to `false` for a element w/ `hidden` which then means the ternary returns `true` right?

**Comment by majornista** [comment]
I've noticed that the tree walker also breaks if there are items that are hidden by ancestors with CSS. You can see this in the second, Toolbar example on the FocusScope page: https://react-spectrum.adobe.com/react-aria/FocusScope.html#usefocusmanager-example. One is unable to tab beyond the example toolbar because the element the tree walker is trying to focus, the hamburger menu button, is hidden by a media query.

**Comment by snowystinger** [comment]
My understanding the with ask about tree walker was to get rid of the scope array. Instead, a tree walker would be created pointing to the first child within the scope, from there, the walker can access siblings and children. This would mean we wouldn't need to keep querying the dom to update our scope list of all focusable items, instead we'd just determine the next focusable item as needed. We can use the sentinels still if we need to.
Let me know if I misunderstood.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 4]
```diff
@@ -119,30 +119,35 @@ export function useFocusManager(): FocusManager {
...
-      if (!previousNode && opts.wrap) {
+      if (!previousNode && wrap) {
         previousNode = focusable[0];
       }
       if (previousNode) {
-        previousNode.focus();
+        focusElement(previousNode, true);
       }
       return previousNode;
     }
```
Comment (majornista) [inline packages/@react-aria/focus/src/FocusScope.tsx:153]: TODO refactor to use tree walking starting from end sentinel, while still accounting for `!!(node.compareDocumentPosition(n) & (Node.DOCUMENT_POSITION_PRECEDING | Node.DOCUMENT_POSITION_CONTAINED_BY))` filter.

Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx]: I have refactored this with cce1045a1f2125173f042c231e9f58cbfc722509.

**Comment by majornista** [comment]
@snowystinger and @LFDanLu, I have finally refactored this to use `getFocusableTreeWalker`, removing need for `getFocusableElementsInScope`.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 5]
```diff
@@ -119,30 +119,46 @@ export function useFocusManager(): FocusManager {
...
-        previousNode = focusable[0];
+      let scope = scopeRef.current;
+      let {from, tabbable, wrap} = opts; 
+      let node = from || document.activeElement;
+      let sentinel = scope[scope.length - 1].nextElementSibling;
+      let walker = getFocusableTreeWalker(getScopeRoot(scope), {tabbable});
+      walker.currentNode = sentinel;
+      let prevNode = walker.previousNode() as HTMLElement;
+      let previousNode;
+      while (prevNode && isElementInScope(prevNode, scope))  {
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:147]: so I tried your branch and replaced this section with the mirror of the above focusNext code, it works in all cases, except for two tests, one in DatePicker and one in DateRangePicker, if I had to guess, those two components need to be fixed, rather than this code needing to compensate for them. It's either that, or FocusScope.test.tsx needs to have a test case that highlights what that different use case is. I'm looking into what's happening in DatePicker now, but if you already know, I figured I'd comment here.

**Review by snowystinger — COMMENTED** [review]
It looks like you've found some edge cases, can we get tests for those? https://github.com/adobe/react-spectrum/pull/1493/commits/cc9656850ff1af0ca58b79932b2839fed67797df

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 5 above)
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:147]: Datepicker has the event hander `onMouseDown` in `useDateField`, which calls `focusManager.focusPrevious` with the dateField element as the `from` prop. The Datepicker test expects the last segment element within the dateField that handles the mouse down event to receive focus.  However the the dateField container passed as the `from` element is not focusable and has no preceding focusable elements, so without calling  `compareDocumentPosition` to locate a segment before or contained within the `from` element, the FocusScope will simply wrap focus to the last item within the scope. For a DateRangePicker, this could be the last segment in a dateField other than the one that triggered `onMouseDown`.

Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:147]: Note that with `focusNext`, I don't think this is a problem, because focusable elements contained within a `from` element will always follow the `from` element when tree walking.

**Comment by majornista** [comment]
@snowystinger I've added unit tests for the edge case, navigating out of a FocusScope and back to an element in the `document.body`, and the `focusPrevious` example that was failing tests in Datepicker.

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 5 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:147]: should DatePicker actually call `focusNext` or maybe call focus on the last field directly instead of using the focusManager? or why must it call `focusPrevious`?

I'm actually playing with the storybook built from main, I'm unable to click into a DatePicker at all...

I guess what i'm getting at is, what does it mean to be calling focusPrevious... is this actually a real use case?

@devongovett ?

Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:147]: I'm not sure that this is a real use case. My PR to address Datepicker issues, https://github.com/adobe/react-spectrum/pull/246, corrects the behavior of onPressStart to set focus to date segments, and in my testing within Storybook, I haven't been able to get the onMouseDown event within useDateField to ever actually get called.

Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:147]: Then my opinion is that we delete the test as it isn't valid, I haven't been able to fire that event either, and we simplify our logic so forward and backward match. Especially since your other PR addresses it. I'll ping devon directly too to make sure he's ok with it.

File: packages/@react-aria/utils/src/isElementVisible.ts (hunk 1 above)
Comment (majornista) [reply packages/@react-aria/utils/src/isElementVisible.ts:49]: I think `isAttributeVisible` will return `false` for `!element.hasAttribute('hidden')` when the element has a `hidden` attribute before ever hitting the ternary, wrapped in parentheses, that follows: `(element.nodeName === 'DETAILS' && childElement && childElement.nodeName !== 'SUMMARY'  ? element.hasAttribute('open') : true)`. The ternary accounts for the edge case of a `<details>` without a `<summary>` that does not have an `open` attribute, the contents of which will not be visible. 

Example: https://codepen.io/majornista/pen/YzpeKRG?editors=1111

**Review by LFDanLu — APPROVED** [review]
LGTM, tested locally with a FocusScope w/ various hidden buttons/inputs and looks like it works properly. Containment and wrapping also look good to me

**Review by LFDanLu — COMMENTED** [review]
Confirmed that the latest commit fixes the example given in https://github.com/adobe/react-spectrum/issues/1570

**Review by snowystinger — COMMENTED** [review]
Datepicker still has issue but they are the same issues it had before. Otherwise, lgtm! once the notice is in place, i'll approve

**Review by snowystinger — DISMISSED** [review]
Thanks!

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 6]
```diff
@@ -119,30 +119,36 @@ export function useFocusManager(): FocusManager {
...
+      let walker = getFocusableTreeWalker(getScopeRoot(scope), {tabbable});
+      walker.currentNode = isElementInScope(node, scope) ? node : sentinel;
+      let nextNode = walker.nextNode() as HTMLElement;
+      if ((!nextNode || !isElementInScope(nextNode, scope)) && wrap) {
+        walker.currentNode = sentinel;
+        nextNode = walker.nextNode() as HTMLElement;
       }
-      if (nextNode) {
-        nextNode.focus();
+      if (nextNode && isElementInScope(nextNode, scope)) {
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:133]: All the `isElementInScope` checks make me wonder if there should be a way to add that into the tree walker filter so we don't have to do this in so many places...

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 7]
```diff
@@ -165,21 +171,13 @@ const focusableElements = [
...
-    if (node.matches(selector)) {
-      res.push(node);
-    }
-    res.push(...Array.from(node.querySelectorAll(selector)));
-  }
-  return res;
+const TABBABLE_ELEMENT_SELECTOR = focusableElements.join(':not([hidden]):not([tabindex="-1"]),');
+
+function getScopeRoot(scope: HTMLElement[]) {
+  return scope.length ? scope[0].parentElement : null;
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:180]: Possibly, I'll have to check. If there are no children in the scope, the `getFocusableTreeWalker` probably shouldn't do anything anyway.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 8]
```diff
@@ -361,11 +354,15 @@ function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boole
         e.preventDefault();
         e.stopPropagation();
         if (nextElement) {
-          nextElement.focus();
+          focusElement(nextElement, true);
         } else {
           // If there is no next element, blur the focused element to move focus to the body.
           focusedElement.blur();
         }
+      } else if (nextElement && !isElementInScope(nextElement, scope)) {
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:362]: It would, but we I'm not sure we can unit test that behavior. I think this was added to fix a unit test failure.

[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:362]: See test at https://github.com/adobe/react-spectrum/pull/1493/files/b9e06919bd772a844edb55d6342696b5940448e5#diff-4ede2be95e8bcfd65b5421b9ae5394aa17e3fc11f77217b9088166f655690007R832-R858

[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-aria/focus/src/FocusScope.tsx:362]: Ah ok. You might be able to try `userEvent.tab()` instead of firing the keyboard event, which should simulate the browser's default behavior more fully.

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:180]: @devongovett I think the case of `getFocusableTreeWalker` breaking when `null` is passed as the root node is a moot point, because in each case where `getScopeRoot` is called, the `scope.length` will have a length:

In `focusNext` returned by `createFocusManager`:
https://github.com/adobe/react-spectrum/blob/b9e06919bd772a844edb55d6342696b5940448e5/packages/%40react-aria/focus/src/FocusScope.tsx#L126

In `focusPrevious` returned by `createFocusManager`:
https://github.com/adobe/react-spectrum/blob/b9e06919bd772a844edb55d6342696b5940448e5/packages/%40react-aria/focus/src/FocusScope.tsx#L143

In `onKeyDown` event handler in `useFocusContainment`, where if the `document.activeElement` is not in scope the tree walker will not be created:
https://github.com/adobe/react-spectrum/blob/b9e06919bd772a844edb55d6342696b5940448e5/packages/%40react-aria/focus/src/FocusScope.tsx#L204

And in `focusFirstInScope`:
https://github.com/adobe/react-spectrum/blob/b9e06919bd772a844edb55d6342696b5940448e5/packages/%40react-aria/focus/src/FocusScope.tsx#L297

I may simplify `getScopeRoot` to simply return `scope[0].parentElement`.

**Comment by tristandubbeld** [comment]
Hey @majornista! I've encountered an issue where tabbing to the next element outside the FocusScope breaks when it's affected by `display: none`. This PR seems to _almost_ fix that issue but there were some issues I ran into while checking out these changes and writing a few tests.

The first issue I've encountered is actually a bug in the `@testing-library/user-event` package, where `userEvent.tab` is focussing elements affected by `display: none`, which is no bueno. It has been fixed in the latest version so upgrading should do the trick, but it seems that it's going to be quite the task as upgrading breaks some other tests.

```
Test Suites: 4 failed, 1 skipped, 127 passed, 131 of 132 total
Tests:       20 failed, 204 skipped, 2818 passed, 3042 total
```

I'm not sure about the upgrade policy but if I can help with upgrading please let me know.

The second issue is that [the following code inside `useRestoreFocus`](https://github.com/adobe/react-spectrum/blob/8571a8083c325efb991128f541e8cba993bd0409/packages/%40react-aria/focus/src/FocusScope.tsx#L340-L342) seems to break focus restoration in my case:

```tsx
if (!document.body.contains(nodeToRestore) || nodeToRestore === document.body) {
  nodeToRestore = null;
}
```

If I remove code above all tests (including the new ones I wrote) seem to succeed. Here's the test I wrote:

```tsx
it('should move focus outside to the next tabbable element past non-tabbable elements', function () {
  let {getByTestId} = render(
    <div>
      <FocusScope>
        <button data-testid="button1">button1</button>
        <button data-testid="button2">button2</button>
      </FocusScope>
      <div style={{display: 'none'}}>
        <button>button none</button>
      </div>
      <button data-testid="button3">button3</button>
    </div>
  );

  let button1 = getByTestId('button1');
  let button2 = getByTestId('button2');
  let button3 = getByTestId('button3');

  act(() => {button1.focus();});
  expect(document.activeElement).toBe(button1);

  userEvent.tab();
  expect(document.activeElement).toBe(button2);

  userEvent.tab();
  expect(document.activeElement).toBe(button3);

  userEvent.tab({shift: true});
  expect(document.activeElement).toBe(button2);
});
```

If it were possible to submit a PR to a PR I would, but I'm not sure if I can so hopefully this comment will do the job. If I need to clarify some things or can help in any other way I'd be glad to!

**Comment by majornista** [comment]
@tristandubbeld Removing  [the following code inside useRestoreFocus](https://github.com/adobe/react-spectrum/blob/8571a8083c325efb991128f541e8cba993bd0409/packages/%40react-aria/focus/src/FocusScope.tsx#L340-L342) breaks the following test at https://github.com/adobe/react-spectrum/pull/1493/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts#diff-4ede2be95e8bcfd65b5421b9ae5394aa17e3fc11f77217b9088166f655690007R833-R859:

```ts
  describe('scope child of document.body', function () {
    it('should navigate in and out of scope in DOM order when the nodeToRestore is the document.body', function () {
      function Test() {
        return (
          <div>
            <input data-testid="beforeScope" />
            <FocusScope>
              <input data-testid="inScope" />
            </FocusScope>
            <input data-testid="afterScope" />
          </div>
        );
      }

      let {getByTestId} = render(<Test />);
      let beforeScope = getByTestId('beforeScope');
      let inScope = getByTestId('inScope');
      let afterScope = getByTestId('afterScope');

      act(() => {inScope.focus();});
      userEvent.tab();
      expect(document.activeElement).toBe(afterScope);
      act(() => {inScope.focus();});
      userEvent.tab({shift: true});
      expect(document.activeElement).toBe(beforeScope);
    });
  });
```

**Comment by majornista** [comment]
@tapananand FocusScope's `getFocusableTreeWalker` should account for the use case in your test, because filters out elements with `display:none` using our own `isElementVisible` method https://github.com/adobe/react-spectrum/pull/1493/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts#diff-71362b654c8ff884f72f97185679bc594b22872637db5f09d8eeba0b6ca551e3R20-R24. I think the test failure is as you say in the `@testing-library/user-event` package `userEvent.tab`. An earlier commit of mine, [explicitly set focus to the next tabbable element in the DOM returned by the focusable tree walker](https://github.com/adobe/react-spectrum/pull/1493/commits/44fcb24aea96e7764b75da22416f6bcdae52f8a2#diff-1b4bb1d7973d81d8ac44ea584d527bfbe742409b0783b2087c06bc356af0365dL362-L366), but I removed that in favor of assuming the default browser behavior would work in the real world and be replicated by `userEvent.tab` in the tests.

**Comment by majornista** [comment]
@tristandubbeld, the [the following code inside useRestoreFocus](https://github.com/adobe/react-spectrum/blob/8571a8083c325efb991128f541e8cba993bd0409/packages/%40react-aria/focus/src/FocusScope.tsx#L340-L342) was changed to include ` || nodeToRestore === document.body`, because when the `nodeToRestore` is the `document.body`, the `useRestoreFocus` method sets the currentNode of the tree walker as the document.body, and focus would be restored to the first element tabbable element in the document.body regardless of where in the DOM the FocusScope exists. 

```ts
      if (!document.body.contains(nodeToRestore) || nodeToRestore === document.body) {
          nodeToRestore = null;
      }

      // If there is no next element, or it is outside the current scope, move focus to the
      // next element after the node to restore to instead.
      if ((!nextElement || !isElementInScope(nextElement, scope)) && nodeToRestore) {
        walker.currentNode = nodeToRestore;
```

**Comment by tristandubbeld** [comment]
Hey @majornista, thanks for the detailed answer! You're right, I've checked again and I might have overlooked some changes when testing. First it was confusing to me that the `nodeToRestore` was the body (thinking it would focus on the first focusable element rather than the next), but it makes sense, thanks for the explanation.

**Comment by majornista** [comment]
@tristandubbeld As far as I can tell, the test case you created is capturing the bug in `@testing-library/user-event` package `userEvent.tab` that should not be reproducible in the real world, where the default behavior of the Tab key should just work. I can update the version of testing-library, but it is breaking 7 tests in NumberField, that all seem to be related to handling input of either non-numeric characters or Arabic text, which seems to point to more bugs in the new implementation of capturing keyboard input with `@testing-library/user-event`. Is there a real-world use case that you can reproduce where these FocusScope updates are not working?

**Comment by tristandubbeld** [comment]
@majornista I created [this CodeSandbox](https://codesandbox.io/s/focus-scope-contain-toggle-with-display-none-3wp2e) a few days ago to reproduce the issue I was having. For this PR I tried reproducing the issue in storybook and it doesn't occur anymore. **This PR fixes the issue**! If there's anything else I can do to help you in getting it merged let me know, happy to help.

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 6 above)
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:133]: I guess we have to distinguish between a tree walker that returns only elements in scope and one that includes the whole DOM, the `focusPrevious/focusNext` use case versus the `onKeyDown/Tab` use case.

Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx]: I've managed to reduce the number of calls to `isElementInScope` with: https://github.com/adobe/react-spectrum/pull/1493/commits/c63c243866630322c0315e0243ceae6ede32f472.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 9]
```diff
@@ -405,14 +398,15 @@ export function getFocusableTreeWalker(root: HTMLElement, opts?: FocusManagerOpt
           return NodeFilter.FILTER_REJECT;
         }
 
-        if ((node as HTMLElement).matches(selector)) {
+        if ((node as HTMLElement).matches(selector)
+          && isElementVisible(node as HTMLElement)
+          && (scope ? isElementInScope(node as HTMLElement, scope) : true)) {
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:403]:
```suggestion
          && (!scope || isElementInScope(node as HTMLElement, scope))) {
```

File: packages/@react-aria/utils/src/isElementVisible.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
+/*
```
Comment (devongovett) [inline packages/@react-aria/utils/src/isElementVisible.ts:1]: IMO we should keep this within the focus package for now since it's only used there

**Review by devongovett — COMMENTED** [review]
Looks great!

**Review by snowystinger — APPROVED** [review]
lgtm!

**Review by devongovett — APPROVED** [review]

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 11]
```diff
@@ -405,14 +399,15 @@ export function getFocusableTreeWalker(root: HTMLElement, opts?: FocusManagerOpt
...
+        if ((node as HTMLElement).matches(selector)
+          && isElementVisible(node as HTMLElement)
+          && (!scope || isElementInScope(node as HTMLElement, scope))) {
           return NodeFilter.FILTER_ACCEPT;
         }
 
         return NodeFilter.FILTER_SKIP;
       }
-    },
-    false
```
Comment (Andarist) [inline packages/@react-aria/focus/src/FocusScope.tsx:415]: I know it's not quite your business as you don't support IE11 but I have just wanted to leave this information here for future reference. Removing this argument has broken compatibility with IE11 as it doesn't treat it as optional.
